### ISSUE TITLE
Provide a library for bitvector-like valued named constants

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5384]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5384|
|Type|enhancement|
|Created|2011-11-13 19:34:38Z|
|Branch|https://github.com/twisted/twisted/tree/flag-constants-5384|

[#4671](https://github.com/twisted/twisted/issues/4671) proposed adding a lot of new symbolic constant features to a helper somewhere in Twisted. This is for the kind of symbolic constant that has a value like a bit in a bitvector and which can be combined from other constants which are part of the same logical group, using all of the traditional bitwise operations.

This would be useful in conch, because SSH uses lots of bitvector-style constants.

<details><summary>Searchable metadata</summary>

```
trac-id__5384 5384
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_flag_constants_5384 branches/flag-constants-5384
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1321212878000000 1321212878000000
changetime__1331456720000000 1331456720000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [33419]) Branching to 'flag-constants-5384'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

I implemented many features.  Possibly there is room for more, but maybe this is a good self-contained start?  One thing that I'm somewhat uncertain about is the behavior of `(SomeFlags.A | SomeFlags.B) is (SomeFlags.A | SomeFlags.B)` or perhaps even more interestingly, `SomeFlags.A is (SomeFlags.A | SomeFlags.B ^ SomeFlags.B)` (in contrast to `SomeFlags.A is SomeFlags.A`).

Also, I feel like the documentation is terribly repetitive.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/flag-constants-5384) 

Finally, this is the last ticket for [#4671](https://github.com/twisted/twisted/issues/4671), so when I close this ticket I expect to close that ticket as well.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

The buildbots are in a pretty nasty state, and the patch is a bit large, but I like what I see so far.  I will think about it some more and hopefully look at a more comprehensive list of build results when the builders are back online.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

 * 
```
+    Tests for L{twisted.python.constdants.Flags}, a base class for containers of
```

Typo: constants.

 * Flags and FlagConstant should be added to `constants.__all__`.

 * 
```
+        result._realize(self._container, set([]), 0)
```

I don't think you need to pass anything to build an empty set.

 * 
```
+        bits set which were not set in the original constant.  constants.
```

There is an extra "constants" here.

 * It may be more interesting to also test READ in FlagConstantNegationTests.test_value, as otherwise you don't check completely the value passed in the realize call in invert.

 * 
```
+            if (flag.value & self.value) == 0:
```

You don't need the parenthesis here.

Great branch, please merge if the buildot is happy.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [33677]) Merge flag-constants-5384 

Author: exarkun
Reviewer: therve
Fixes: [#5384](https://github.com/twisted/twisted/issues/5384), [#4671](https://github.com/twisted/twisted/issues/4671)

Add bitvector-like constant support to `twisted.python.constants`.
